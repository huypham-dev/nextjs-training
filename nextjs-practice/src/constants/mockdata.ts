export const CATEGORIES = [
  {
    id: "1",
    name: "mobile",
    image:
      "https://images.prismic.io/nextjsexample/7853f863-6a12-4312-a6da-f930437f6d62_mobile.png",
  },
  {
    id: "2",
    name: "internet",
    image:
      "https://images.prismic.io/nextjsexample/cb71ce70-bb03-442b-99e8-a318255b81f3_3579187.png",
  },
];

export const ARTICLES = [
  {
    id: "1",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
  {
    id: "2",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
];
