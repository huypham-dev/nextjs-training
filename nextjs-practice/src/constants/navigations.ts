export const navigations = [
  { label: "Home", path: "/" },
  { label: "Blog", path: "/blog" },
  { label: "Contact", path: "/contact" },
];
