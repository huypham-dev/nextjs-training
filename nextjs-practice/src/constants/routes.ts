export const ROUTES = {
  home: {
    path: "/",
  },
  blog: {
    path: "/blog",
    withId: (id: string) => `/blog/${id}`,
  },
  post: {
    path: "/post",
    withId: (id: string) => `/post/${id}`,
  },
  contact: {
    path: "/",
  },
};
