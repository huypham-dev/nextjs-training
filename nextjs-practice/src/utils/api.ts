const BASE_URL = "https://5f7c244700bd74001690a4a7.mockapi.io";
export const COLLECTION = {
  CATEGORIES: "/categories",
  POSTS: "/posts",
};

export const API = {
  fetchAllCategories: async () => {
    const res = await fetch(`${BASE_URL}${COLLECTION.CATEGORIES}`);

    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchPostsByCategory: async (categoryId) => {
    const res = await fetch(
      `${BASE_URL}${COLLECTION.POSTS}?category=${categoryId}`
    );

    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchAllPost: async () => {
    const res = await fetch(`${BASE_URL}${COLLECTION.POSTS}`);
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchPostItem: async (id) => {
    const res = await fetch(`${BASE_URL}${COLLECTION.POSTS}/${id}`);
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },

  fetchFeaturedPosts: async () => {
    const res = await fetch(
      `${BASE_URL}${COLLECTION.POSTS}?sortBy=view&order=desc&p=1&l=6`
    );
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
};
