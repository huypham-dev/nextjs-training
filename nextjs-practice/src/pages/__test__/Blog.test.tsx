import { mount, shallow } from "enzyme";
import Blog, { getStaticPaths, getStaticProps } from "../blog/[[...path]]";
import { mockArticles, mockCategories } from "constants/mockdata";
import { API } from "utils/api";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require("next/router"), "useRouter");
const fetchPostsByCategory = jest.spyOn(API, "fetchPostsByCategory");
const fetchAllPost = jest.spyOn(API, "fetchAllPost");
const fetchAllCategories = jest.spyOn(API, "fetchAllCategories");
const paths = [
  { params: { path: ["mobile"] } },
  { params: { path: ["internet"] } },
  { params: { path: [] } },
];

beforeEach(() => {
  useRouter.mockImplementationOnce(() => ({
    query: { path: ["mobile"] },
  }));
});

describe("Blog page", () => {
  it("Snapshot", async () => {
    const container = shallow(
      <Blog categories={mockCategories} posts={mockArticles} />
    );

    expect(container).toMatchSnapshot();
  });

  it("Render page correctly with props", () => {
    const container = mount(
      <Blog categories={mockCategories} posts={mockArticles} />
    );

    expect(container.prop("categories")).toEqual(mockCategories);
    expect(container.prop("posts")).toEqual(mockArticles);
  });

  it("getStaticProps should return props with all posts", async () => {
    fetchAllPost.mockImplementationOnce(() => Promise.resolve(mockArticles));
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve(mockCategories)
    );
    const response = await getStaticProps({ params: { path: null } });

    expect(response).toEqual(
      expect.objectContaining({
        props: {
          posts: mockArticles,
          categories: mockCategories,
          error: mockArticles,
        },
      })
    );
  });

  it("getStaticProps should return props with posts by category", async () => {
    fetchPostsByCategory.mockImplementationOnce(() =>
      Promise.resolve(mockArticles)
    );
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve(mockCategories)
    );
    const response = await getStaticProps({ params: { path: "mobile" } });

    expect(response).toEqual(
      expect.objectContaining({
        props: {
          posts: mockArticles,
          categories: mockCategories,
          error: mockArticles,
        },
      })
    );
  });

  it("getStaticPaths should return props with fallback is false", async () => {
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve(mockCategories)
    );
    const response = await getStaticPaths({});

    expect(response).toEqual(
      expect.objectContaining({
        paths,
        fallback: false,
      })
    );
  });

  it("getStaticPaths should return props with fallback is true", async () => {
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve({
        errorCode: 404,
        message: "Not found",
      })
    );
    const response = await getStaticPaths({});

    expect(response).toEqual(
      expect.objectContaining({
        paths: [{ params: { path: [] } }],
        fallback: true,
      })
    );
  });
});
