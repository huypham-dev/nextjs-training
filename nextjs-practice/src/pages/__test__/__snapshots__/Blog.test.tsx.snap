// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog page Snapshot 1`] = `
<div
  className="mx-auto max-w-6xl bg-white py-4 px-12 shadow-xl mb-24"
>
  <Head>
    <title>
       
      Mobile
       | Blog
    </title>
  </Head>
  <Memo(CategoryTitle)
    category="mobile"
  />
  <div
    className="container max-w-6xl mx-auto md:flex items-start py-8 md:px-0"
  >
    <Memo(Articles)
      data={
        Array [
          Object {
            "author": "author 1",
            "category": "internet",
            "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
            "createdAt": "2021-03-01T12:59:16.855Z",
            "id": "1",
            "image": "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
            "slug": "slug 1",
            "title": "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
            "view": 90,
          },
          Object {
            "author": "author 1",
            "category": "internet",
            "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
            "createdAt": "2021-03-01T12:59:16.855Z",
            "id": "2",
            "image": "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
            "slug": "slug 1",
            "title": "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
            "view": 90,
          },
        ]
      }
    />
    <Memo(CategoriesWidget)
      data={
        Array [
          Object {
            "id": "1",
            "image": "https://images.prismic.io/nextjsexample/7853f863-6a12-4312-a6da-f930437f6d62_mobile.png",
            "name": "mobile",
          },
          Object {
            "id": "2",
            "image": "https://images.prismic.io/nextjsexample/cb71ce70-bb03-442b-99e8-a318255b81f3_3579187.png",
            "name": "internet",
          },
        ]
      }
    />
  </div>
</div>
`;
